CEO-232 Add unsaved changes warning

rjsheperd opened the pull request:
## Purpose
<!-- Description of what has been added/changed -->
Show a unsaved changes warning when a user tries to hit "Back" and has unsaved answers.

## Related Issues
Closes CEO-232

## Submission Checklist
- [x] Commits include the JIRA issue and the `#review` hashtag (e.g. `CEO-### #review <comment>`)
- [x] Code passes linter rules (`npm run eslint`/`clj-kondo --lint src`)

## Testing
<!-- Create a BDD style test script -->
1. Given I am a user, When I am collecting, And I have unsaved answers, And I try to navigate away, Then I am shown a unsaved changes warning.

## Screenshots
<!-- Add a screen shot when UI changes are included -->
#### Unsaved changes alert
<img width="1418" alt="Screen Shot 2021-09-29 at 2 20 22 PM" src="https://user-images.githubusercontent.com/1829313/135350130-eea54984-0891-48b0-a1ca-168108cc7cb1.png">

#### Unsaved confirm alert when using the plot nav
![image](https://user-images.githubusercontent.com/1829313/135532343-34a59808-3b17-4732-9430-18ddda15a03d.png)

sirmspencer commented:
src/js/collection.js
56, instead of maintaining a separate state just check if any questions have been answered.
324, 325, can the events just be called once in didMount and didUnmount?
Check on quit click, change plot click, and beforeunload.

rjsheperd commented:
Update with:
* Removed `isUnsaved` and now use `_.isEqual` to compare the original/new sample values
* Added check for `hasChanged` to the plot navigation functions (see screenshots above)
* Moved the add/remove listener fn to componentDidMount/WillUnmount

sirmspencer commented:
src/js/collection.js
323, 445, Move your functions to the same place in the code. Above confirmUnsaved looks good.
367, If you are creating a helper function for repeatability, include the check for !this.hasChanged()
385, confirmUnsaved does not take any input.
620, 655, this.setState is async so in this form, you cant guarantee that the value is reset.  Do this synchronously by passing "ignoreCheck" into `navToNextPlot`.  This is more clear what you are trying to do anyway.

sirmspencer commented:
src/js/collection.js
375, `ignoreCheck === true` is redundant `if (ignoreCheck || this.confirmUnsaved())`
